### PR TITLE
show store location topology

### LINKED ISSUE #688: Visualize Location Topology
## Feature Request

**Is your feature request related to a problem? Please describe:**

Users may mis-configure location labels. This problem can be easily discovered if we provide a UI to visualize the topology.

**Describe the feature you'd like:**

Add a new web page, which visualizes store location labels.

Labels can be visualized using [Tree-Graph](https://observablehq.com/@d3/collapsible-tree) as:

image

Different colors can be assigned to better distinguish from labels and stores.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

Notes:

1. Stores have different kind of labels. Besides location labels, there could also be label `engine = tiflash`, which need to be filtered out.

2. Starting from PD v4.0.0, PD supports Placement Rules. When Placement rules are specified, the location labels configuration from PD might not be reliable since it can be overridden by placement rules. To resolve this issue, we may fetch all placement rules and check whether location label orders in the placement rules conflicts with the rules defined in PD configuration. If there is a conflict, the visualization should stop.

### PR DESCRIPTION
**Pull request by baurine**
resolve #688 

API response example:

```json
{
  "location_labels": "zone,rack,host",
  "stores": [
    {
      "address": "127.0.0.1:20160",
      "labels": {
        "host": "h1",
        "rack": "r1",
        "zone": "sh"
      }
    },
    {
      "address": "127.0.0.1:20161",
      "labels": {
        "host": "h2",
        "rack": "r1",
        "zone": "sh"
      }
    },
    {
      "address": "127.0.0.1:20162",
      "labels": {
        "host": "h1",
        "rack": "r1",
        "zone": "bj"
      }
    },
    {
      "address": "127.0.0.1:3930",
      "labels": {
        "engine": "tiflash",
        "zone": "bj"
      }
    }
  ]
}
```

Aggregate to tree nodes:

```json
{
  "name": "zone",
  "value": "",
  "children": [
    {
      "name": "zone",
      "value": "sh",
      "children": [
        {
          "name": "rack",
          "value": "r1",
          "children": [
            {
              "name": "host",
              "value": "h1",
              "children": [
                {
                  "name": "address",
                  "value": "127.0.0.1:20160",
                  "children": []
                }
              ]
            },
            {
              "name": "host",
              "value": "h2",
              "children": [
                {
                  "name": "address",
                  "value": "127.0.0.1:20161",
                  "children": []
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "name": "zone",
      "value": "bj",
      "children": [
        {
          "name": "rack",
          "value": "r1",
          "children": [
            {
              "name": "host",
              "value": "h1",
              "children": [
                {
                  "name": "address",
                  "value": "127.0.0.1:20162",
                  "children": []
                }
              ]
            }
          ]
        },
        {
          "name": "address",
          "value": "127.0.0.1:3930",
          "children": []
        }
      ]
    }
  ]
}
```

Currently, it is displayed in a new tab in the cluster-info page:

image

### CODE REVIEW

File: ui/lib/apps/ClusterInfo/components/StoreLocation.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import client, { TopologyStoreLocation } from '@lib/client'
+import { Pre } from '@lib/components'
+
+type TreeNode = {
+  name: string
+  value: string
+  children: TreeNode[]
+}
+
+function buildTopology(data: TopologyStoreLocation | undefined) {
```
Comment (baurine) [inline ui/lib/apps/ClusterInfo/components/StoreLocation.tsx:12]: It is easier to do the aggregation in the frontend, so I didn't do it in the backend.

**Comment by baurine** [comment]
Hi, @breeswish @HunDunDM please help take a look, thanks! the UI component to show the tree is WIP by the intern.

File: pkg/utils/topology/models.go [hunk 2]
```diff
@@ -57,6 +57,16 @@ type StoreInfo struct {
 	StartTimestamp int64             `json:"start_timestamp"`
 }
 
+type StoreLabels struct {
+	Address string            `json:"address"`
+	Labels  map[string]string `json:"labels"`
+}
+
+type StoreLocation struct {
+	LocationLabels string        `json:"location_labels"`
```
Comment (breeswish) [inline pkg/utils/topology/models.go:66]: maybe it is better to be an array

File: ui/lib/apps/ClusterInfo/components/StoreLocation.tsx (hunk 1 above)
Comment (breeswish) [reply ui/lib/apps/ClusterInfo/components/StoreLocation.tsx:12]: I agree

File: pkg/utils/topology/models.go (hunk 2 above)
Comment (baurine) [reply pkg/utils/topology/models.go:66]: ok, will update is soon.

**Comment by baurine** [comment]
Current effect:

image

TODO:
- [ ] Use d3 to render instead of echarts

**Comment by breeswish** [comment]
Good job! I suggest to assign different colors to distinguish a Label and an Instance.

**Comment by baurine** [comment]
> Good job! I suggest to assign different colors to distinguish a Label and an Instance.

Got it!

Final effect:



@breeswish PTAL, thanks!

**Comment by baurine** [comment]
Changed colors for label and instance:

image

**Comment by breeswish** [comment]
Very cool!

**Comment by baurine** [comment]
How to build the test cluster: see `etc/manualTestEnv/multiReplica/README.md`

**Pull request merged**